rmw_microxrcedds_c/src/rmw_node_names.c

```diff
@@ -14,18 +14,88 @@
 
 #include <rmw/rmw.h>
 #include <rmw/error_handling.h>
+#include <rmw/sanity_checks.h>
+
+#include <rmw_microxrcedds_c/config.h>
+#include <rmw_microxrcedds_c/rmw_c_macros.h>
+
+#include <rcutils/types/string_array.h>
+
+#include "./types.h"
+#include "./identifiers.h"
+
+#ifdef RMW_UXRCE_GRAPH
+#include "./rmw_graph.h"
+#endif  // RMW_UXRCE_GRAPH
 
 rmw_ret_t
 rmw_get_node_names(
   const rmw_node_t * node,
   rcutils_string_array_t * node_names,
   rcutils_string_array_t * node_namespaces)
 {
+#ifdef RMW_UXRCE_GRAPH
+  // Perform RMW checks
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(node, node->implementation_identifier,
+    eprosima_microxrcedds_identifier, return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  if (RMW_RET_OK != rmw_check_zero_rmw_string_array(node_names)) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_RET_OK != rmw_check_zero_rmw_string_array(node_namespaces)) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  // Get micro_ros_msgs/msg/Graph instance
+  rmw_uxrce_node_t * custom_node = (rmw_uxrce_node_t *)(node->data);
+  rmw_graph_info_t * graph_info = &custom_node->context->graph_info;
+
+  if (!graph_info->initialized) {
+    return RMW_RET_OK;
+  }
+
+  rmw_ret_t ret = RMW_RET_OK;
+  micro_ros_msgs__msg__Graph * graph_data = micro_ros_msgs__msg__Graph__create();
+
+  if (RMW_RET_OK != rmw_graph_fill_data_from_buffer(graph_info, graph_data)) {
+    ret = RMW_RET_ERROR;
+    goto fini;
+  }
+
+  // Init node name and namespaces string array
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  if (RCUTILS_RET_OK != rcutils_string_array_init(
+      node_names, graph_data->nodes.size, &allocator)) {
+    ret = RMW_RET_ERROR;
+    goto fini;
+  }
+  if (RCUTILS_RET_OK != rcutils_string_array_init(
+      node_namespaces, graph_data->nodes.size, &allocator)) {
+    ret = RMW_RET_ERROR;
+    goto fini;
+  }
+
+  // Copy information into result arrays
+  for (size_t i = 0; i < graph_data->nodes.size; ++i) {
+    micro_ros_msgs__msg__Node * node = &graph_data->nodes.data[i];
+    node_namespaces->data[i] = allocator.zero_allocate(
+      strlen(node->node_namespace.data) + 1, sizeof(char), allocator.state);
+    strcpy(node_namespaces->data[i], node->node_namespace.data);
+    node_names->data[i] = allocator.zero_allocate(
+      strlen(node->node_name.data) + 1, sizeof(char), allocator.state);
+    strcpy(node_names->data[i], node->node_name.data);
+  }
+
+fini:
+  micro_ros_msgs__msg__Graph__destroy(graph_data);
+  return ret;
```

Comment:
Returning ok when error.
```suggestion
  return RMW_RET_ERROR;
```

Reply:
Same as above. I could've done something like
```
    ...
    micro_ros_msgs__msg__Graph__destroy(graph_data);
    return RMW_RET_OK;
fini:
    micro_ros_msgs__msg__Graph__destroy(graph_data);
    return RMW_RET_ERROR;
```
but it is kind of redundant for me.

Reply:
Cool, closing